xkb/xkb.c

```diff
@@ -4521,57 +4521,41 @@ XkbSizeGeomShapes(XkbGeometryPtr geom)
     return size;
 }
 
-static char *
-XkbWriteGeomShapes(char *wire, XkbGeometryPtr geom, Bool swap)
+static void XkbWriteGeomShapes(x_rpcbuf_t *rpcbuf, XkbGeometryPtr geom)
 {
     int i;
     XkbShapePtr shape;
-    xkbShapeWireDesc *shapeWire;
 
     for (i = 0, shape = geom->shapes; i < geom->num_shapes; i++, shape++) {
         register int o;
         XkbOutlinePtr ol;
-        xkbOutlineWireDesc *olWire;
 
-        shapeWire = (xkbShapeWireDesc *) wire;
-        shapeWire->name = shape->name;
-        shapeWire->nOutlines = shape->num_outlines;
-        if (shape->primary != NULL)
-            shapeWire->primaryNdx = XkbOutlineIndex(shape, shape->primary);
-        else
-            shapeWire->primaryNdx = XkbNoShape;
-        if (shape->approx != NULL)
-            shapeWire->approxNdx = XkbOutlineIndex(shape, shape->approx);
-        else
-            shapeWire->approxNdx = XkbNoShape;
-        shapeWire->pad = 0;
-        if (swap) {
-            swapl(&shapeWire->name);
-        }
-        wire = (char *) &shapeWire[1];
+        /* write xkbShapeWireDesc */
+        x_rpcbuf_write_CARD32(rpcbuf, shape->name);
+        x_rpcbuf_write_CARD8(rpcbuf, shape->num_outlines);
+        x_rpcbuf_write_CARD8(
+            rpcbuf,
+            shape->primary ? XkbOutlineIndex(shape, shape->primary) : XkbNoShape);
+        x_rpcbuf_write_CARD8(rpcbuf,
+            shape->approx ? XkbOutlineIndex(shape, shape->approx) : XkbNoShape);
+        x_rpcbuf_write_CARD8(rpcbuf, 0); /* pad1 */
+
         for (o = 0, ol = shape->outlines; o < shape->num_outlines; o++, ol++) {
             register int p;
             XkbPointPtr pt;
-            xkbPointWireDesc *ptWire;
-
-            olWire = (xkbOutlineWireDesc *) wire;
-            olWire->nPoints = ol->num_points;
-            olWire->cornerRadius = ol->corner_radius;
-            olWire->pad = 0;
-            wire = (char *) &olWire[1];
-            ptWire = (xkbPointWireDesc *) wire;
+
+            /* write xkbOutlineWireDesc */
+            x_rpcbuf_write_CARD8(rpcbuf, ol->num_points);
+            x_rpcbuf_write_CARD8(rpcbuf, ol->corner_radius);
+            x_rpcbuf_pad(rpcbuf);
```

Comment:
Not sure, maybe an explicit `x_rpcbuf_write_CARD16` would be cleaner?

Reply:
Yep, in some other places you write a field like `pad` explicitly. The result is the same I guess.